### PR TITLE
add support for X.509 key usage attributes (code signing and others)

### PR DESCRIPTION
**Pull request by jluebbe**
X.509 certificates can have (extended) key usage attributes to restrict their usage (or "purpose" in OpenSSL). This is mainly useful when signing RAUC bundles in not the only use-case for a PKI.

This PR contains some minor cleanups before adding a `check-purpose` option in the `[keyring]` section of `system.conf`. As OpenSSL does not have a predefined purpose value corresponding to the extendedKeyUsage=codeSinging attribute, we need to register a custom purpose check callback (`codesign`).

### CODE REVIEW

**Comment by jluebbe** [comment]
@Emantor I've added some more details to the docs.

File: docs/advanced.rst [hunk 1]
```diff
@@ -44,15 +44,57 @@ exceeds the scope of this documentation.
...
+  basicConstraints = CA:FALSE
+  extendedKeyUsage=critical,codeSigning
+
+As OpenSSL does not (yet) provide a purpose check for code signing, RAUC
+contains its own implementation, which can be enabled with the
+:ref:`check-purpose=codesign <check-purpose>` configuration option.
+For the leaf (signer) certificate, the *extendedKeyUsage* attribute must exist
+and contain (at least) the *codeSigning* value.
+Also, if it has the *keyUsage* attribute, it must contain at least *digitalSignature*.
+For all other (issuer) certificates the chain, the *extendedKeyUsage* attribute
```
Comment (ejoerns) [inline docs/advanced.rst:88]: *in* the chain is meant here I assume

File: docs/advanced.rst [hunk 2]
```diff
@@ -44,15 +44,57 @@ exceeds the scope of this documentation.
...
 
+.. _sec-key-usage:
+
+Certificate Key Usage Attributes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default (for backwards compatibility reasons), RAUC does not check the
+certificate's key usage attributes.
+When not using a stand-alone PKI for RAUC, it can be useful to enable checking
+via the ``check-usage`` configuration option to allow only specific
```
Comment (ejoerns) [inline docs/advanced.rst:64]: s/check-usage/check-purpose/

File: docs/advanced.rst [hunk 3]
```diff
@@ -44,15 +44,57 @@ exceeds the scope of this documentation.
...
+certificates for bundle signing.
```
Comment (ejoerns) [inline docs/advanced.rst:65]: Is 'signing' or 'installation' meant here?

File: docs/advanced.rst (hunk 1 above)
Comment (jluebbe) [reply docs/advanced.rst:88]: Yes, fixed.

File: docs/advanced.rst (hunk 3 above)
Comment (jluebbe) [reply docs/advanced.rst:65]: Yes, 'installation' is clearer here.

File: docs/advanced.rst (hunk 2 above)
Comment (jluebbe) [reply docs/advanced.rst:64]: Fixed.

**Comment by Emantor** [comment]
:partying_face: :tada:

**Pull request merged**